cpp/src/search/contains_table.cu

```diff
@@ -173,124 +194,98 @@ rmm::device_uvector<bool> contains(table_view const& haystack,
+    preprocessed_haystack, preprocessed_needles);
 
   // The output vector.
   auto contained = rmm::device_uvector<bool>(needles.num_rows(), stream, mr);
 
-  auto const preprocessed_needles =
-    cudf::experimental::row::equality::preprocessed_table::create(needles, stream);
-  // Check existence for each row of the needles table in the haystack table.
-  {
-    auto const needles_it = cudf::detail::make_counting_transform_iterator(
-      size_type{0}, [] __device__(auto const idx) { return rhs_index_type{idx}; });
-
-    auto const hasher = cudf::experimental::row::hash::row_hasher(preprocessed_needles);
-    auto const d_hasher =
-      strong_index_hasher_adapter{hasher.device_hasher(nullate::DYNAMIC{has_any_nulls})};
-
-    auto const comparator = cudf::experimental::row::equality::two_table_comparator(
-      preprocessed_haystack, preprocessed_needles);
-
-    auto const check_contains = [&](auto const value_comp) {
-      if (cudf::detail::has_nested_columns(haystack) or cudf::detail::has_nested_columns(needles)) {
-        auto const d_eqcomp =
-          comparator.equal_to<true>(nullate::DYNAMIC{has_any_nulls}, compare_nulls, value_comp);
-        map.contains(needles_it,
-                     needles_it + needles.num_rows(),
-                     contained.begin(),
-                     d_hasher,
-                     d_eqcomp,
-                     stream.value());
-      } else {
-        auto const d_eqcomp =
-          comparator.equal_to<false>(nullate::DYNAMIC{has_any_nulls}, compare_nulls, value_comp);
-        map.contains(needles_it,
-                     needles_it + needles.num_rows(),
-                     contained.begin(),
-                     d_hasher,
-                     d_eqcomp,
-                     stream.value());
-      }
-    };
-
-    dispatch_nan_comparator(compare_nans, check_contains);
+  auto const haystack_iter = cudf::detail::make_counting_transform_iterator(
+    size_type{0}, [] __device__(auto idx) { return lhs_index_type{idx}; });
+  auto const needles_iter = cudf::detail::make_counting_transform_iterator(
+    size_type{0}, [] __device__(auto idx) { return rhs_index_type{idx}; });
+
+  auto const helper_func = [&](auto const& d_self_equal, auto const& d_two_table_equal) {
+    auto const d_equal = comparator_adapter{d_self_equal, d_two_table_equal};
+
+    auto set = cuco::experimental::static_set{
+      cuco::experimental::extent{compute_hash_table_size(haystack.num_rows())},
+      cuco::empty_key{lhs_index_type{-1}},
+      d_equal,
+      cuco::experimental::linear_probing<1, hasher_type>{d_hasher},
+      detail::hash_table_allocator_type{default_allocator<lhs_index_type>{}, stream},
+      stream.value()};
+
+    if (haystack_has_nulls && compare_nulls == null_equality::UNEQUAL) {
+      auto const row_bitmask =
+        cudf::detail::bitmask_and(haystack, stream, rmm::mr::get_current_device_resource()).first;
```

Comment:
`bitmask_and` only exec `and` of the top level columns in the table. If we have nested columns with nulls nested inside, this `row_bitmask` will be all valid. What we really want is to exclude any row having any nested nulls from being inserted into the set.

Reply:
Imagine that we have a structs column `struct<int, int>` where the second `int` child is all nulls. In such situation, we will hit the performance issue reported in the issues at lines 203-204 (deleted) above.

Reply:
There is still one more potential situation that we need to explore/verify: `nulls` at top level are considered unequal but nested nulls are always equal. In such situation, we must update our row comparator to support that before we can use `bitmask_and` here.

Reply:
> What we really want is to exclude any row having any nested nulls from being inserted into the set.

Is this the desired behavior of spark? If so, we should probably move `build_row_bitmask` to a detail header thus it can be used in other places like join or distinct, etc. Also, would null equality make a difference here? 

For my own education, there are several details I would like to confirm:

- If nulls are unequal, and two tables are not nullable at the top level, Are the below rows considered equal?
```
list{ list{ list{ 1, null } } }  v.s.  list{ list{ list{ 1, null } } }
``` 
- The same condition as above, but the top-level lists are nullable, would this make a difference?
- I vaguely recall for certain stream compaction operations, Spark treats all nulls as equal


> nulls at top level are considered unequal but nested nulls are always equal.

Maybe I'm missing something obvious, do we distinct top-level nulls and nested nulls in libcudf?

Reply:
>> What we really want is to exclude any row having any nested nulls from being inserted into the set.

> Is this the desired behavior of spark? 
This is not asked by Spark. This is due to the performance issue reported in the deleted lines 203-204.

Indeed, in spark, in some cases, nested nulls are considered **equal** even top-level nulls are **unequal**. However, we currently cannot compare nulls with both equal and unequal in cudf.

In summary, when nulls are unequal, we should avoid inserting rows having any nulls at any nested level, for performance reason.

Reply:
Done.